### PR TITLE
refactor permissions and groups

### PR DESCRIPTION
**Pull request by xuwhite**
based on:
Screenshot 2024-04-20 at 22 31 35

Legend 
red - delete
green - insert
yellow - do something else

### CODE REVIEW

File: includes/pages/admin_arrive.php [hunk 1]
```diff
@@ -206,15 +209,17 @@ function admin_arrive()
...
+            ['name' => __('general.name'),],
+            ($admin_arrive ? ['rendered_planned_arrival_date' => __('Planned arrival')] : []),
+            ['arrived' => __('Arrived?')],
+            ($admin_arrive ? [
+                'rendered_arrival_date' => __('Arrival date'),
+                'rendered_planned_departure_date' => __('Planned departure'),
+                'actions' => __('general.actions'),
+            ] : [])
+        ), $users_matched),
+        div('row', $admin_arrive ?? [
```
Comment (MyIgel) [inline includes/pages/admin_arrive.php:222]: That prints a 1 at the end instead of the statistics

File: includes/pages/admin_arrive.php [hunk 2]
```diff
@@ -19,48 +19,51 @@ function admin_arrive()
...
     $search = '';
     $request = request();
+    $admin_arrive = auth()->can('admin_arrive');
 
     if ($request->has('search')) {
         $search = strip_request_item('search');
         $search = trim($search);
     }
 
     $action = $request->get('action');
```
Comment (MyIgel) [inline includes/pages/admin_arrive.php:29]: Can too be evaluated inside the if

File: src/Middleware/LegacyMiddleware.php [hunk 3]
```diff
@@ -51,10 +51,14 @@ public function process(
             $page = str_replace('-', '_', $page);
         }
 
+        $arriveList = false;
```
Comment (MyIgel) [inline src/Middleware/LegacyMiddleware.php:54]: I would name it somewhat more generic like `$pageAllowed` to allow for further permission overwrites on legacy pages

File: db/migrations/2024_04_20_000000_refactor_permissions_and_groups.php [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    protected string $shifttypesView = 'shifttypes.view';
+
+    protected Connection $db;
+
+    public function __construct(SchemaBuilder $schema)
+    {
+        parent::__construct($schema);
+        $this->db = $this->schema->getConnection();
+
+        $this->active = $this->getPrivilegeId('admin_active');
```
Comment (MyIgel) [inline db/migrations/2024_04_20_000000_refactor_permissions_and_groups.php:48]: Having a mapping of `'permission.name' => 'id'` would make the code below imho more readable but thats an option too

**Review by MyIgel — CHANGES_REQUESTED** [review]

**Comment by MyIgel** [comment]
Other todos from above list:
* Bureaucrat & ShiCo admin_log
* Bureaucrat are missing the schedule.import but Devs have it now?

File: includes/pages/admin_arrive.php (hunk 2 above)
Comment (xuwhite) [reply includes/pages/admin_arrive.php:29]: -v

Comment (xuwhite) [reply includes/pages/admin_arrive.php:29]: yes I could add it to both if and else if

Comment (xuwhite) [reply includes/pages/admin_arrive.php:29]: move down would've helped me more

File: src/Middleware/LegacyMiddleware.php [hunk 5]
```diff
@@ -51,10 +51,14 @@ public function process(
...
+        if ($page === 'admin_arrive') {
+            $arriveList = $this->auth->can('users.arrive.list');
```
[the remark this answers is not in the document]
Comment (xuwhite) [reply src/Middleware/LegacyMiddleware.php]: please implement it

File: includes/pages/admin_arrive.php (hunk 2 above)
Comment (MyIgel) [reply includes/pages/admin_arrive.php:29]: I meant the `action` part but yeah

**Pull request merged**